python/dgl/sampling/pinsage.py

```diff
@@ -0,0 +1,212 @@
+    graph where the neighbors of each node are the most commonly visited nodes of the
+    same type by random walk with restarts.  The random walk with restarts are based
+    on a given metapath, which should have the same beginning and ending node type.
+
+    The homogeneous graph also has a feature that stores the number of visits to
+    the corresponding neighbors from the seed nodes.
+
+    Parameters
+    ----------
+    G : DGLHeteroGraph
+        The heterogeneous graph.
+    random_walk_length : int
+        The maximum number of steps of random walk with restarts.
+
+        Note that here we consider a full traversal of the given metapath as a single
+        random walk "step" (i.e. a single step may consist of multiple hops).
+
+        Usually considered a hyperparameter.
+    random_walk_restart_prob : int
+        Restart probability of random walk with restarts.
+
+        Note that the random walks only would halt after a full traversal of a metapath.
+        It will never halt in the middle of a metapath.
+
+        Usually considered a hyperparameter.
+    num_random_walks : int
+        Number of random walks to try for each seed node.
+
+        Usually considered a hyperparameter.
+    num_neighbors : int
+        Number of neighbors to select for each seed.
+    metapath : list[str] or list[tuple[str, str, str]], optional
+        The metapath.
+
+        If not given, assumes that the graph is homogeneous.
+    weight_column : str, default "weights"
+        The weight of each neighbor, stored as an edge feature.
+
+    Inputs
+    ------
+    seed_nodes : Tensor
+        A tensor of seed node IDs of node type ``ntype``.
+
+    Outputs
+    -------
+    g : DGLHeteroGraph
+        A homogeneous graph constructed by selecting neighbors for each seed node according
+        to PinSAGE algorithm.
+
+    Examples
+    --------
+    See examples in :any:`PinSAGESampler`.
+    """
+    def __init__(self, G, random_walk_length, random_walk_restart_prob,
+                 num_random_walks, num_neighbors, metapath=None, weight_column='weights'):
+        self.G = G
+        self.weight_column = weight_column
+        self.num_random_walks = num_random_walks
+        self.num_neighbors = num_neighbors
+        self.random_walk_length = random_walk_length
+
+        if metapath is None:
+            if len(G.ntypes) > 1 or len(G.etypes) > 1:
+                raise ValueError('Metapath must be specified if the graph is homogeneous.')
+            metapath = [G.canonical_etypes[0]]
+        start_ntype = G.to_canonical_etype(metapath[0])[0]
+        end_ntype = G.to_canonical_etype(metapath[-1])[-1]
+        if start_ntype != end_ntype:
+            raise ValueError('The metapath must start and end at the same node type.')
+        self.ntype = start_ntype
+
+        self.metapath_hops = len(metapath)
+        self.metapath = metapath * random_walk_length
```

Comment:
Probably would save both the `metapath` for each step and a `full_metapath = metapath * random_walk_length`.